### ISSUE TITLE
[Data] Enabling local shuffle buffer reduces throughput when iterating with Ray Trainer

### ISSUE THREAD

**Issue by scottjlee** [ray-project/ray#42317]
### What happened + What you expected to happen

When iterating over a Ray Dataset within the `TorchTrainer` train loop, a non-`None` `local_shuffle_buffer_size` causes a decrease in throughput compared to disabling the local shuffle buffer. 

In the case with 1 GPU node with the `multi_node_train_benchmark`, we observed 417 img/s (local shuffle buffer enabled) vs 753 img/s (disabled).

### Versions / Dependencies

Ray master

### Reproduction script

Any `multi_node_train_benchmark` related benchmark with prefetch batches specified and empty model, e.g.
```
python multi_node_train_benchmark.py --num-workers 1 --file-type image --use-gpu --num-epochs 2 --skip-train-model --prefetch-batches 16
```

### Issue Severity

None

**Comment by oceanusxiv** [comment]
@scottjlee Hi, I'm curious has there been any updates on this issue? I've seen this severely impact our training loop also.

**Comment by ae20cg** [comment]
@scottjlee following up on this - my perf for tabular data loading (arrays of 12K float32s) goes from about 200s for 1 epoch to 1200s for 1 epoch when I set a `local_shuffle_buffer_size` to <= batch_size * prefetch_batches.

this seems like a major degradation. are there any tips for improving this (outside of the doc page that already exists)?

**Comment by scottjlee** [comment]
Sorry for the delay on this folks, we haven't had bandwidth to look into this code path for quite some time.

Other than the tips on [this docs page](https://docs.ray.io/en/latest/data/shuffling-data.html), I would suggest:
- upgrading to latest version of ray
- using file-based shuffling when possible, as this is the most efficient
- lowering `local_shuffle_buffer_size` until the "Batch iteration time breakdown" time shown in `ds.stats()` is low

**Comment by ae20cg** [comment]
@scottjlee 

no problem. i am using file based shuffling

what about if you perform a `map_batches(shuffle_func, batch_size=None)` on the dataset? is it possible you could get better shuffle performance - assuming the block size is very different from the batch_size leading to some differentiation in batches epoch/epoch?

**Comment by apacha** [comment]
My two cents here after spending a day, trying to eliminate the bottleneck in my training pipeline: I have the feeling that the local_shuffle_buffer negates the prefetch_batches logic, which obviously causes slowdowns.

Claude analyzed the code and gave this explanation (I haven't dived into the code to confirm):

> Why it stalls: iter_batches runs a single background thread that does: prefetch → ray.get() (serial) → ShufflingBatcher (pure sync Python). The ShufflingBatcher blocks all output until the buffer accumulates 1.5 ×  shuffle_buffer_size rows for an O(N) in-place shuffle. During that entire fill, the formatting threadpool (where prefetch_batches has effect) sits idle. The prefetch threads are starved, not disabled. So prefetch_batches=N literally does nothing useful when local_shuffle_buffer_size is set.

**Comment by richardliaw** [comment]
@xinyuangui2  recently did a deep dive into this issue.

TLDR - if you're seeing slowdowns, you should move the shuffle into a `map_batches` call. 

We took 2 independent paths to improve window size:

1. Improving `iter_batches(local_shuffle_buffer)`  
2. Moving the shuffle into a map\_batches task.

**Improving iter\_batches**

As mentioned above, existing `iter_batches(local_shuffle_buffer)` exhibited performance issues and also would OOM frequently. In our experiments, we were able to reproduce throughput degradation and OOM issues.

We found that the cause of this were two-fold:

1. Existing implementation would reshuffle all the data in the buffer   
2. The frequency of refilling the data was very frequent (it would buffer 1.5x the stated size, add data and reshuffle every time the buffer hit 1.0x the stated size)  


To address these, we:

1. Decreased the shuffle frequency to only trigger when the buffer size hit 0.5x the total buffer  
2. Shuffled the data indices (instead of the full data) and incrementally sampled the data from the buffer.

### Incremental vs Baseline

Below is the time needed for a shuffle operation on different windows of 128MB blocks.

| Window size | Time for shuffling |
| :---- | :---- |
| 1GB | 1.44s |
| 4GB | 5.78s |
| 16GB | 22.89s |

We noticed that the shuffling was done on a different thread, and it was not necessary to move all the data at once in a shuffle. Instead, we could pipeline the shuffling by shuffling the indices and fetching the shuffled rows incrementally so that it would always overlap with the training.

To observe the improvement from incrementally indexing the dataset, we evaluated the improvement of simply taking a subset of rows compared to the original implementation:

| Buffer Rows | Buffer GB | take(all) | take(4096) | Speed Ratio | Size Ratio (N/4096) |
| :---- | :---- | :---- | :---- | :---- | :---- |
| 500,000 | 0.04G | 0.0292s | 0.000354s | 82x | 122x |
| 2,000,000 | 0.16G | 0.2047s | 0.000491s | 417x | 488x |
| 6,250,000 | 0.50G | 1.2414s | 0.000596s | 2,082x | 1,526x |
| 25,000,000 | 2.00G | 6.9811s | 0.000916s | 7,623x | 6,104x |
| 50,000,000 | 4.00G | 15.1349s | 0.001003s | 15,091x | 12,207x |
| 100,000,000 | 8.00G | 31.2443s | 0.001048s | 29,818x | 24,414x |

### Improvements on baseline

On the image classification dataset, which:

* has 513 files, each file with 93.4MB  
* On 4 training workers  
* 1 x g4dn.12xlarge(48 vCPU, 4 NVIDIA T4 GPUs, 192 GiB memory)

|  | Baseline | Incremental Indexing | Baseline | Incremental Indexing |
| :---- | :---- | :---- | :---- | :---- |
| **Shuffle Threshold** | **1.0x** | **1.0x** | **0.5x** | **0.5x** |
| 0 GB | 437 |  |  |  |
| 1 GB | 426 | 438 | 421 | 443 |
| 2 GB | OOM | 391 | 431 | 452 |
| 3 GB | OOM | OOM | OOM | 432 |
| 4 GB | OOM | OOM | OOM | OOM |

**We see that the incremental indexing is able to maintain a higher throughput for a larger threshold, but still sees OOMs at small buffer sizes.**

On the Recsys dataset:

* has 2120 files, each file with 167.8MB  
* On 8 training workers  
* 1 x p4d.24xlarge (96 vCPU, 8 NVIDIA A100 40GB GPUs, 1152 GiB memory)

|  | Baseline | Incremental Indexing | Baseline | Incremental Indexing |
| :---- | :---- | :---- | :---- | :---- |
| **Shuffle Threshold** | **1.0x** | **1.0x** | **0.5x** | **0.5x** |
| 0 GB | 919k |  |  |  |
| 1 GB | 854k | 833k | 925k | 934k |
| 2 GB | 900k | 878k | 917k | 928k |
| 3 GB | 653k | 754k | 792k | 923k |
| 4 GB | OOM | 592k | OOM | 678k |

**We see that the incremental indexing is able to maintain a higher throughput for a larger threshold as well.**

Changing the reshuffle threshold may cause the shuffle quality to degrade since less unique files appear in a batch. As we varied the window size, we noticed that there was marginal degradation of shuffle quality when decreasing the shuffle threshold.

| Reshuffle threshold | Image (1gb) | Recsys (1gb) |
| :---- | ----- | ----- |
| 1.0 \* window\_size | 14.6 | 185 |
| 0.75 \* window\_size | 13.2 | 184 |
| 0.5 \* window\_size | 13.3 | 180 |
| 0.25 \* window\_size | 12.4 | 181 |
| 0.01 \* window\_size | 9.4 | 176 |

*Shuffle quality relative degradation.*   
*Baseline is measured by how many average unique shuffle units (files) are in each batch.*  

### map\_batches (offloading the shuffle)

We also implemented an offloaded shuffle by using map\_batches. 

Shuffle throughput for image workload

| buffer size | Baseline | Incremental \+ 0.5x threshold | map\_batches (1gpu) | map\_batches (1 gpu \+ 1 cpu) |
| :---- | :---- | :---- | :---- | :---- |
| 0G | 437 |  |  |  |
| 1G | 426 | 443 | 438 | 429 |
| 2G | OOM | 452 | 441 | 429 |
| 3G | OOM | 432 | 296 | 427 |
| 4G | OOM | OOM | 269 | 420 |

Shuffle throughput for recsys workload

| buffer size | Baseline | Incremental \+ 0.5x threshold | map\_batches (1gpu) | map\_batches (1 gpu \+ 1 cpu) |
| :---- | :---- | :---- | :---- | :---- |
| 0 GB | 919k |  |  |  |
| 1 GB | 854k | 934k | 915k | 1.002M |
| 2GB | 900k | 928k | 911k | 976K |
| 3GB | 653k | 923k | 887k | 983K |
| 4 GB | OOM | 678k | 542k | 1.090M |

**By offloading the shuffle onto the CPU, we saw the throughput maintained across different buffer sizes.**

## These changes landed
* Added the [user guide](https://docs.ray.io/en/master/data/shuffling-data.html#map-batches-shuffle) for map_batches shuffle
* Landed the local shuffle [incremental index-taking](https://github.com/ray-project/ray/commit/bfa2f1cf19401cac741ff2163052709431133295) optimization, reducing memory usage by 2.5–3x and improving throughput by up to 2–3x.

Note, the shuffle buffer approach will still cause performance degradation at certain buffer sizes, so best would be to move that off into a separate map-batches shuffle operator if possible.
